### DISCUSSION TITLE
Best way to get the path that a car took during a lap

### DISCUSSION THREAD

**Discussion by DaviMargo** [theOehrly/Fast-F1#356] [Q&A] [answered]
Hello everyone, my goal with this API is to use it inside a 3D Software to generate a path.

This is my code comparing 2 laps of the 2023 Australian GP, im trying to generate an array with all the coordinates of the car during the lap but the starting points of the 2 laps isn't correct.

Also comparing the 2 path with live footage the distances between the 2 cars during the lap dont seem correct either, can anyone help me with that? @theOehrly

`import bpy
import numpy as np
import fastf1

fastf1.core.Telemetry.TELEMETRY_FREQUENCY = 100
session = fastf1.get_session(2023, 'Australia', 'Q')
session.load()
driver1 = session.laps.pick_driver('VER').pick_fastest()
driver2 = session.laps.pick_driver('RUS').pick_fastest()

def animateCar(driver, obj):
    
    x = driver.telemetry['X']
    y = driver.telemetry['Y']
    z = driver.telemetry['Z']

    # Convert to array
    x = x.to_numpy()
    y = y.to_numpy()
    z = z.to_numpy()
    
    t = 0
    positionsPath = []

    for index, i in enumerate(x):
        if t < len(x):
            coordinate = [x[t]/, y[t]/, 0] # don't need Z axis
            t = t + 1
            positionsPath.append(coordinate)
            
 animateCar(driver1, 'VER')`
            
 How do i get the exact same start of the lap of the 2 cars (basically when they are both on the start/finish line)

**Comment by theOehrly** [comment]
Hey, so a few things are likely problematic here.

1. The data is not synchronized perfectly. This is a problem of the source data and FastF1 can only try to reduce that problem to some amount.
2. The sample rate is very low at ~200ms. Therefore, you will almost never have a sample that aligns perfectly with the start finish line.
3. You should think of the position data as "normalized" coordinates. They are (more or less) normalized to one path, so you cannot extract different cornering lines, for example.

**Reply by DaviMargo** [reply]
Thanks a lot for answering!
Is there anything i can do to improve that apart from fixing the positions manually?

Also, i've seen that i can get the cars GPS positions with the get_pos_data function, would that be better in any Way?

Do you know of any other way to get the correct car positions on the track?

**Reply by theOehrly** [reply]
> Thanks a lot for answering! Is there anything i can do to improve that apart from fixing the positions manually?

If you figure out how to align the data correctly, tell me 😅 I've been trying to reliably improve that for a long time.

> Also, i've seen that i can get the cars GPS positions with the get_pos_data function, would that be better in any Way?

Oh, yes. I didn't notice that you are using `.telemetry`. The data there is interpolated. `.get_pos_data` will give slightly better results.

`.telemetry` merges car data and position data. Those have different timebases, therefore this process involves interpolation which is not ideal for accuracy.

You should read
https://theoehrly.github.io/Fast-F1/core.html#fastf1.core.Lap.get_telemetry
https://theoehrly.github.io/Fast-F1/howto_accurate_calculations.html

> Do you know of any other way to get the correct car positions on the track?

No

**Reply by theOehrly** [reply, marked as answer]
I don't know whether they are using FastF1. I'd say it is likely. But in any case, if they don't have insider information/contacts, then they are using the exact same raw data. Even if they don't download it through FastF1.
